okhttp/src/jvmMain/kotlin/okhttp3/Dispatcher.kt

```diff
@@ -164,12 +196,27 @@ class Dispatcher() {
     val executableCalls = mutableListOf<AsyncCall>()
     val isRunning: Boolean
     synchronized(this) {
-      val i = readyAsyncCalls.iterator()
-      while (i.hasNext()) {
-        val asyncCall = i.next()
+      while (readyAsyncCalls.isNotEmpty()) {
+        val filteredCalls = readyAsyncCalls.mapNotNull {
+          if (it.callsPerHost.get() < maxRequestsPerHost) it.call else null
+        }
+
+        if (filteredCalls.isEmpty()) break // Max host capacity for all requests.
+
+        val nextCall = asyncCallDispatcher.nextCall(filteredCalls)
+
+        val i = readyAsyncCalls.iterator()
+
+        var asyncCall: AsyncCall = i.next()
+
+        while (asyncCall.call != nextCall && i.hasNext()) {
```

Comment:
This looks accidentally quadratic. We should avoid this second pass.  

Not sure what the answer is, allowing it? a more optimised filteredCalls? state on AsyncCall to check?

Reply:
I think quadratic runtime here is not a big deal because the outer loop runs until either `readyAsyncCalls` is empty or `runningAsyncCalls.size >= maxRequests`. `maxRequests` is 64 by default so even in the worst case the outer loop will only run 64 times.

Also I think the run time added by this check is fairly negligible compared to the network response time.

Reply:
Also I think achieving linear time is impossible because we have to convert `AsyncCall` to `Call` anyways unless you are okay with exposing `AsyncCall`.